trie/triedb/pathdb/history.go

```diff
@@ -0,0 +1,594 @@
+
+	for addrHash, slots := range states.Storages {
+		for slotHash := range slots {
+			storageList[addrHash] = append(storageList[addrHash], slotHash)
+		}
+		slices.SortFunc(storageList[addrHash], func(a, b common.Hash) bool { return a.Less(b) })
+	}
+	for hash := range states.Incomplete {
+		incomplete = append(incomplete, hash)
+	}
+	slices.SortFunc(incomplete, func(a, b common.Hash) bool { return a.Less(b) })
+
+	return &history{
+		meta: &meta{
+			version:     stateHistoryVersion,
+			parent:      parent,
+			root:        root,
+			blockNumber: blockNumber,
+			incomplete:  incomplete,
+		},
+		accounts:    states.Accounts,
+		accountList: accountList,
+		storages:    states.Storages,
+		storageList: storageList,
+	}
+}
+
+// encode serializes the state history and returns four byte streams represent
+// concatenated account/storage data, account/storage indexes respectively.
+func (h *history) encode() ([]byte, []byte, []byte, []byte) {
+	var (
+		slotNumber     uint32 // the number of processed slots
+		accountData    []byte // the buffer for concatenated account data
+		storageData    []byte // the buffer for concatenated storage data
+		accountIndexes []byte // the buffer for concatenated account index
+		storageIndexes []byte // the buffer for concatenated storage index
+	)
+	for _, addrHash := range h.accountList {
+		aIndex := accountIndex{
+			Hash:   addrHash,
+			Length: uint8(len(h.accounts[addrHash])),
+			Offset: uint32(len(accountData)),
+		}
+		slots, exist := h.storages[addrHash]
+		if exist {
+			// Encode storage slots in order
+			for _, slotHash := range h.storageList[addrHash] {
+				sIndex := slotIndex{
+					Hash:   slotHash,
+					Length: uint8(len(h.storages[addrHash][slotHash])),
+					Offset: uint32(len(storageData)),
+				}
+				storageData = append(storageData, h.storages[addrHash][slotHash]...)
+				storageIndexes = append(storageIndexes, sIndex.encode()...)
+			}
+			// Fill up the storage meta in account index
+			aIndex.StorageOffset = slotNumber
+			aIndex.StorageNumber = uint32(len(slots))
+			slotNumber += uint32(len(slots))
+		}
+		accountData = append(accountData, h.accounts[addrHash]...)
+		accountIndexes = append(accountIndexes, aIndex.encode()...)
+	}
+	return accountData, storageData, accountIndexes, storageIndexes
+}
+
+// decoder wraps the byte streams for decoding with extra meta fields.
+type decoder struct {
+	accountData    []byte // the buffer for concatenated account data
+	storageData    []byte // the buffer for concatenated storage data
+	accountIndexes []byte // the buffer for concatenated account index
+	storageIndexes []byte // the buffer for concatenated storage index
+
+	lastAccount     common.Hash // the address hash of last resolved account
+	lastAccountRead uint32      // the read-cursor position of account data
+	lastSIndexRead  uint32      // the read-cursor position of storage index
+	lastStorageRead uint32      // the read-cursor position of storage data
```

Comment:
Perhaps lets call the previous one lastSlotIndexRead and this one lastSlotDataRead? It seems cleaner and kind of similar in length anyway.